### PR TITLE
fix(secretsmanager): `arnForPolicies` evaluates to the partial ARN if accessed from a cross-env stack

### PR DESCRIPTION
**Pull request by gshpychka**
Currently, `Secret.secretFullArn` returns the partial ARN if the secret is referenced between cross-env stacks.
An obvious practical implication is that `grant*` methods will produce an incorrect ARN for the IAM policies, since a full ARN is required for cross-environment access.

This PR partially fixes the issue - I reimplemented `arnForPolicies` to be lazily evaluated. It checks if the value is being accessed across environments and adds `-??????` to the ARN if it is.

Now, this does not solve the underlying issue of `secretFullArn` returning the partial ARN. While it should return undefined, we have to check how the prop is accessed (same environment or cross-env) before we know whether to return the ARN or `undefined`. If we use a `Lazy` here, it still cannot return `undefined` (only `any` as the closest thing).

So I don't think the underlying cause could be solved currently, that's why I opted for this partial fix that would address the most practical consequence of the bug.
 
This is a partial fix for https://github.com/aws/aws-cdk/issues/22468.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by gshpychka** [comment]
Not sure how to go about an integration test - I've written one but then realized that the AWS account number has to be provided (and included in the snapshots) - what should I do in this case?

**Comment by gshpychka** [comment]
Hm, the integ tests failed because
```
      {
       "Action": [
        "secretsmanager:DescribeSecret",
        "secretsmanager:GetSecretValue"
       ],
       "Effect": "Allow",
       "Principal": {
        "AWS": {
         "Fn::Join": [
          "",
          [
           "arn:",
           {
            "Ref": "AWS::Partition"
           },
           ":iam::",
           {
            "Ref": "AWS::AccountId"
           },
           ":root"
          ]
         ]
        },
        "Service": "ecs-tasks.amazonaws.com"
       },
       "Resource": {
        "Ref": "testsecretAttachment19AD251F"
       }
      }
```

changed to 
```
      {
       "Action": [
        "secretsmanager:DescribeSecret",
        "secretsmanager:GetSecretValue"
       ],
       "Effect": "Allow",
       "Principal": {
        "AWS": {
         "Fn::Join": [
          "",
          [
           "arn:",
           {
            "Ref": "AWS::Partition"
           },
           ":iam::",
           {
            "Ref": "AWS::AccountId"
           },
           ":root"
          ]
         ]
        }
       },
       "Resource": {
        "Ref": "testsecretAttachment19AD251F"
       }
      },
      {
       "Action": [
        "secretsmanager:DescribeSecret",
        "secretsmanager:GetSecretValue"
       ],
       "Effect": "Allow",
       "Principal": {
        "Service": "ecs-tasks.amazonaws.com"
       },
       "Resource": {
        "Ref": "testsecretAttachment19AD251F"
       }
      }
```

Any idea why this could be?

File: packages/aws-cdk-lib/aws-secretsmanager/lib/secret.ts [hunk 1]
```diff
@@ -450,7 +450,18 @@ abstract class SecretBase extends Resource implements ISecret {
...
+        const consumingStack = Stack.of(context.scope);
+        if (this.stack.account !== consumingStack.account ||
+          (this.stack.region !== consumingStack.region &&
+            !consumingStack._crossRegionReferences) || !this.secretFullArn) {
+          return `${this.secretArn}-??????`;
+        } else {
+          return this.secretFullArn;
+        }
+      },
+    });
```
Comment (mrgrain) [inline packages/aws-cdk-lib/aws-secretsmanager/lib/secret.ts:464]: think this is the right approach to solve this 👍🏻

**Comment by mrgrain** [comment]
> testsecretAttachment19AD251F

**Comment by mrgrain** [comment]
(Sorry I accidently closed this)

**Comment by mrgrain** [comment]
> Hm, the integ tests failed because

Which test is this?

> Any idea why this could be?

I suspect the policy deduplication mechanism (search for `mergeStatements`) doesn't think these are combinable anymore. Probably because it is now a lazy token object which are not going to be equal to each other.

**Comment by gshpychka** [comment]
> Which test is this?

It is [aws-rds/test/integ.serverless-cluster-secret-rotation-custom-names](https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk-testing/framework-integ/test/aws-rds/test/integ.serverless-cluster-secret-rotation-custom-names.ts) and [aws-rds/test/integ.serverless-cluster-secret-rotation
](https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk-testing/framework-integ/test/aws-rds/test/integ.serverless-cluster-secret-rotation.ts)

This would make it a breaing change, though.

Regarding an integration test for the fix itself - I'm not really comfortable with making it exempt, but not sure how a env-specific integ test can possibly be implemented.

**Comment by mrgrain** [comment]
> > Which test is this?
> 
> It is [aws-rds/test/integ.serverless-cluster-secret-rotation-custom-names](https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk-testing/framework-integ/test/aws-rds/test/integ.serverless-cluster-secret-rotation-custom-names.ts) and [aws-rds/test/integ.serverless-cluster-secret-rotation ](https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk-testing/framework-integ/test/aws-rds/test/integ.serverless-cluster-secret-rotation.ts)
> 
> This would make it a breaing change, though.
> 
> Regarding an integration test for the fix itself - I'm not really comfortable with making it exempt, but not sure how a env-specific integ test can possibly be implemented.

Back to square one then.

> Currently, Secret.secretFullArn returns the partial ARN if the secret is referenced between cross-env stacks.

Have you figured out _why_ this happens?

**Comment by gshpychka** [comment]
> Have you figured out _why_ this happens?

I believe it's because the default implementation of `secretFullArn` defined in `BaseSecret` returns `this.secretArn` and that doesn't change when a secret is used across envs. It seems that this is only overridden in `.fromSecret*` methods.

**Comment by gshpychka** [comment]
> > Have you figured out _why_ this happens?
> 
> I believe it's because the default implementation of `secretFullArn` defined in `BaseSecret` returns `this.secretArn` and that doesn't change when a secret is used across envs. It seems that this is only overridden in `.fromSecret*` methods.And `secretArn` [uses `Resource.getResourceArnAttribute`](https://github.com/aws/aws-cdk/blob/4f23619ac773e002c48a6bcf51bfb5d890bd2afc/packages/aws-cdk-lib/aws-secretsmanager/lib/secret.ts#L655C4-L655C4) which in across-env case [returns a formatted ARN ](https://github.com/aws/aws-cdk/blob/43013d04f3a2637b799b9a57916751c3d04b9a2f/packages/aws-cdk-lib/core/lib/resource.ts#L296) and `ref` otherwise.

We could change `secretFullArn` to use a custom function instead of `Resource.getResourceArnAttribute` and return something else in case of cross-env access, but what do we return?

**Comment by mrgrain** [comment]
> We could change secretFullArn to use a custom function instead of Resource.getResourceArnAttribute and return something else in case of cross-env access, but what do we return?

Hm, so I guess that makes `secretFullArn` broken whenever we have a cross stack reference. CR would be one way, could we also deal with this special case using import/exports or a SSM parameter? For both we should have some code doing something similar already.

**Comment by gshpychka** [comment]
> Hm, so I guess that makes `secretFullArn` broken whenever we have a cross stack reference.

No, it works fine with cross-stack references if in the same environment (using Cfn imports under the hood). The relevant logic for cross-stack references is implemented elsewhere.

**Comment by mrgrain** [comment]
> No, it works fine with cross-stack references if in the same environment (using Cfn imports under the hood). The relevant logic for cross-stack references is implemented elsewhere.

Yeah, but could we use the same/similar logic for cross-env?

**Comment by gshpychka** [comment]
> > No, it works fine with cross-stack references if in the same environment (using Cfn imports under the hood). The relevant logic for cross-stack references is implemented elsewhere.
> 
> Yeah, but could we use the same/similar logic for cross-env?

Well, Cfn imports don't work cross-env. The cross-env logic is just "resolve the references at synth time".

What do you think we should be focusing on (I'm getting a bit lost in the discussion)?
Do you think the IAM policy merging could be acceptable? Because whatever we do, seems that there's no way around this.

**Comment by gshpychka** [comment]
> > 
> 
> Yeah, but could we use the same/similar logic for cross-env?

To make sure we're on the same page - detecting that it's a cross-env reference is not the issue - that's easy. Returning a correct `arnForPolicies` in this case is easy as well - this PR does that.

We can have `secretFullArn` return a Lazy as well - and then we can implement the reference with a CR. 

But the issue with merging in IAM policies would still be there in both of these cases, as far as I understand.

**Comment by mrgrain** [comment]
> > > 
> > 
> > 
> > Yeah, but could we use the same/similar logic for cross-env?
> 
> To make sure we're on the same page - detecting that it's a cross-env reference is not the issue - that's easy. Returning a correct `arnForPolicies` in this case is easy as well - this PR does that.
> 
> We can have `secretFullArn` return a Lazy as well - and then we can implement the reference with a CR.
> 
> But the issue with merging in IAM policies would still be there in both of these cases, as far as I understand.

I'm with you now. Having `arnForPolicies` a Lazy is probably okay then. I'll need to set aside a bit to look into the merging code if there is a way around it.

**Comment by mrgrain** [comment]
@gshpychka I can't figure out how to push to your branch. But here's the change

https://github.com/aws/aws-cdk/compare/pr/26308?expand=1

**Review by mrgrain — APPROVED** [review]

**Comment by rv2673** [comment]
@gshpychka @mrgrain This unfortunately break my app. 
However secrets Constructs created through fromSecretCompleteArn, that are used cross region, break completely with this change, since    it is treating them as a partialArn. 

I believe this is because the condition `!this.secretFullArn ` is used in or condition and is never reached since the cross region condition is short circuiting before secretFullArn is actually checked. So even if full arn is filled it is treated as partialArn.

So either the if block should only be entered if fullArn is undefined and cross-region or arnForPolicies should be overriden in class returned fromSecretAttributes.

This is a nice fix for secrets created(not imported) cross-region which would reduce some complexity in my apps since it has this case too.

**Comment by mrgrain** [comment]
@rv2673 Thanks for letting us know! Would you mind creating a new issue for this and include a code sample for the reproduction. Thank you.

**Comment by rv2673** [comment]
> @rv2673 Thanks for letting us know! Would you mind creating a new issue for this and include a code sample for the reproduction. Thank you.

Done https://github.com/aws/aws-cdk/issues/26811

**Pull request merged**